File: Backtracking/39_CombinationSum.py
class Solution(object):

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        lst = []
        candidates.sort()
        self.backtract(lst, [], 0, candidates, target, 0)
        return lst

    def backtract(self, lst, tmplist, list_sum, nums, target, start):
        if list_sum == target:
            lst.append(copy.deepcopy(tmplist))
        else:
            for x in range(start, len(nums)):
                tmplist.append(nums[x])
                list_sum = sum(tmplist)
                if list_sum <= target:
                    self.backtract(lst, tmplist, list_sum, nums, target, x)
                tmplist.pop()

File: Backtracking/39_CombinationSum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> lst = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(lst, new ArrayList<>(), candidates, 0, target, 0);
        return lst;
    }

    private void backtrack(List<List<Integer>> lst, List<Integer> tmplist, int[] nums, int sum, int target, int start) {
        if (sum == target) {
            lst.add(new ArrayList<Integer>(tmplist));
        } else {
            for (int i=start; i<nums.length; i++) {
                tmplist.add(nums[i]);
                sum = tmplist.stream().mapToInt(Integer::intValue).sum();
                if (sum <= target) {
                    backtrack(lst, tmplist, nums, sum, target, i);
                }
                tmplist.remove(tmplist.size() - 1);
            }
        }
    }
}